File: lab_containers/6/src/walk_from_size.cpp
#include <iostream>
#include <fstream>
#include <vector>
#include <forward_list>
#include <list>
#include <set>
#include <map>
#include <iterator>
#include <chrono>

double get_time() {
    return std::chrono::duration_cast<std::chrono::microseconds>
        (std::chrono::steady_clock::now().time_since_epoch()).count()/1e3;
}

template <typename T>
void walk(T& cont)
{
    for (auto c : cont)
    {
        c += 1;
    }
}

void walk(std::map<int, int>& cont){
    for(unsigned int i = 0; i < cont.size(); i++)
    {
        cont[i] += 1;
    }
}


int main()
{
    std::vector<int> v;
    std::forward_list<int> fl;
    std::list<int> l;
    std::set<int> s;
    std::map<int, int> m;

    std::ofstream f1("vector_size.csv", std::ios::out);
    std::ofstream f2("vector_time.csv", std::ios::out);

    std::ofstream g1("forward_list_size.csv", std::ios::out);
    std::ofstream g2("forward_list_time.csv", std::ios::out);

    std::ofstream u1("list_size.csv", std::ios::out);
    std::ofstream u2("list_time.csv", std::ios::out);

    std::ofstream h1("set_size.csv", std::ios::out);
    std::ofstream h2("set_time.csv", std::ios::out);

    std::ofstream v1("map_size.csv", std::ios::out);
    std::ofstream v2("map_time.csv", std::ios::out);

    // Задаём количество элементов
    int n = 100000;

    // Заполняем вектор
    for (int j = 100; j <= n; j += 100)
    {
        for (int i = 0; i < j; i++)
        {
            v.push_back(i);
            fl.push_front(i);
            l.push_front(i);
            s.insert(i);
            m.insert({i, i});
        }
        f1 << v.size() << std::endl;
        g1 << j << std::endl;
        u1 << j << std::endl;
        h1 << s.size() << std::endl;
        v1 << m.size() << std::endl;

        double start;

        //subset
        start = get_time();
        walk(v);
        f2 << get_time() - start << std::endl;

        //set
        start = get_time();
        walk(fl);
        g2 << get_time() - start << std::endl;

        //map
        start = get_time();
        walk(l);
        u2 << get_time() - start << std::endl;

        //multiset
        start = get_time();
        walk(s);
        h2 << get_time() - start << std::endl;

        //multimap
        start = get_time();
        walk(m);
        v2 << get_time() - start << std::endl;

        //Clearing
        v.clear();
        fl.clear();
        l.clear();
        s.clear();
        m.clear();
    }
    return 0;
}

File: lab_containers/6/src/walk_from_size.py
import matplotlib.pyplot as plt
import numpy as np
import csv

with open('vector_size.csv') as f1:
    r1 = csv.reader(f1)
    x1 = []
    for line in r1:
          x1.extend(map(float, line))
with open('vector_time.csv') as f1:
    r1 = csv.reader(f1)
    y1 = []
    for line in r1:
          y1.extend(map(float, line))
with open('forward_list_size.csv') as f1:
    r1 = csv.reader(f1)
    x2 = []
    for line in r1:
          x2.extend(map(float, line))
with open('forward_list_time.csv') as f1:
    r1 = csv.reader(f1)
    y2 = []
    for line in r1:
          y2.extend(map(float, line))
with open('list_size.csv') as f1:
    r1 = csv.reader(f1)
    x3 = []
    for line in r1:
          x3.extend(map(float, line))
with open('list_time.csv') as f1:
    r1 = csv.reader(f1)
    y3 = []
    for line in r1:
          y3.extend(map(float, line))
with open('set_size.csv') as f1:
    r1 = csv.reader(f1)
    x4 = []
    for line in r1:
          x4.extend(map(float, line))
with open('set_time.csv') as f1:
    r1 = csv.reader(f1)
    y4 = []
    for line in r1:
          y4.extend(map(float, line))
with open('map_size.csv') as f1:
    r1 = csv.reader(f1)
    x5 = []
    for line in r1:
          x5.extend(map(float, line))
with open('map_time.csv') as f1:
    r1 = csv.reader(f1)
    y5 = []
    for line in r1:
          y5.extend(map(float, line))

plt.plot(x1, y1, 'b', label = "vector")
plt.plot(x2, y2, 'r', label = "forward_list")
plt.plot(x3, y3, 'g', label = "list")
plt.plot(x4, y4, 'y', label = "set")
plt.plot(x5, y5, 'k', label = "map")
plt.xlabel("size")
plt.ylabel("time, ms")
plt.title("Time of walk from size")
plt.minorticks_on()
plt.grid(which='minor', linestyle=':')
plt.tight_layout()
plt.legend()
plt.savefig("walk_comparison_size_from_10_to_100000_step_100.png")
plt.show()